### PR TITLE
fix(client): update invalid stored code

### PR DESCRIPTION
**Pull request by ojeytonwilliams**
If `loadCodeEpic` encounters stored code it cannot parse, it would silently throw an error.  This fixes old stored code and then clears out anything it doesn't understand.

### CODE REVIEW

File: client/src/templates/Challenges/redux/code-storage-epic.js [hunk 1]
```diff
@@ -150,7 +150,13 @@ function loadCodeEpic(action$, state$) {
             // This was pushed to production with https://github.com/freeCodeCamp/freeCodeCamp/pull/43023
             foundChallengeFile = codeFound[challengeFile.fileKey];
           }
-          const isCodeFound = Object.keys(foundChallengeFile).length > 0;
+          let isCodeFound;
```
Comment (ShaunSHamilton) [inline client/src/templates/Challenges/redux/code-storage-epic.js:153]:
```suggestion
          let isCodeFound = false;
```

**Review by ShaunSHamilton — CHANGES_REQUESTED** [review]
Nit-picky, because the fewer types the more betterer.

File: client/src/templates/Challenges/redux/code-storage-epic.js [hunk 2]
```diff
@@ -150,7 +150,13 @@ function loadCodeEpic(action$, state$) {
             // This was pushed to production with https://github.com/freeCodeCamp/freeCodeCamp/pull/43023
             foundChallengeFile = codeFound[challengeFile.fileKey];
```
Comment (ShaunSHamilton) [inline client/src/templates/Challenges/redux/code-storage-epic.js:151]: Seeing as we are just purging unwanted formats, why not remove this "old code-storage format" stuffs too?

**Comment by ojeytonwilliams** [comment]
@raisedadead what do you think about purging the saved javascript challenge code? Just the challenge code, mind, nothing else.

**Comment by raisedadead** [comment]
> what do you think about purging the saved javascript challenge code?

Does this include user code too?

**Comment by ShaunSHamilton** [comment]
> > what do you think about purging the saved javascript challenge code?
> 
> Does this include user code too?

Yes. This is the localStorage for non-certification projects. So, only user code. Only client-side. (the code saved with `Ctrl + S` in the editor)

**Comment by ShaunSHamilton** [comment]
@raisedadead We discussed and have decided to go with the "hacky" approach which allows _all_ currently saved code.

This is mostly because there are probably a few Campers relying on localStorage for the JS certification projects (before they submit)

**Comment by ojeytonwilliams** [comment]
Blocked while we discuss what we want to implement

**Comment by ojeytonwilliams** [comment]
This now restores the old behaviour, but should gracefully handle old saved code and only discard stored code if it isn't valid.

@Sembauke this should help a little with the mobile app.

**Comment by Sembauke** [comment]
Thanks @ojeytonwilliams ,

I have tried to use the editor on mobile but came across some issues when switching panes and trying to run tests.
(Most of the time the code would just not save)

Also some users report unable to remove text or type at all. 
It is hard to know if these issue's are from the mobile app or from the Monaco Editor itself.

Hope this fixes the storing issues.

**Comment by ojeytonwilliams** [comment]
@Sembauke Monaco really isn't great for mobile.  If you're interested, you can try https://microsoft.github.io/monaco-editor/playground.html on mobile, to see what's a Monaco specific issue rather than something we have done.

This PR is a relatively minor improvement - I just remembered that inability to save was a source of frustration for some mobile users.

**Comment by ShaunSHamilton** [comment]
@Sembauke For reference, we have always pointed mobile users to: https://www.freecodecamp.org/news/freecodecamp-mobile/

**Review by ShaunSHamilton — DISMISSED** [review]
This LGTM 👍 

One thing I noticed, even when updating the old code format, the old fields remain behind:
Is there any reason this would be a problem?

```
{
    "fileKey": "scriptjs",
    "head": "",
    "tail": "",
    "editableRegionBoundaries": null,
    "history": [
        "script.js"
    ],
    "name": "script",
    "ext": "js",
    "path": "script.js",
    "contents": "// The global variable\nvar watchList = [\n  {\n    \"Title\": \"Inception\",\n    \"Year\": \"2010\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"16 Jul 2010\",\n    \"Runtime\": \"148 min\",\n    \"Genre\": \"Action, Adventure, Crime\",\n    \"Director\": \"Christopher Nolan\",\n    \"Writer\": \"Christopher Nolan\",\n    \"Actors\": \"Leonardo DiCaprio, Joseph Gordon-Levitt, Elliot Page, Tom Hardy\",\n    \"Plot\": \"A thief, who steals corporate secrets through use of dream-sharing technology, is given the inverse task of planting an idea into the mind of a CEO.\",\n    \"Language\": \"English, Japanese, French\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Won 4 Oscars. Another 143 wins & 198 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BMjAxMzY3NjcxNF5BMl5BanBnXkFtZTcwNTI5OTM0Mw@@._V1_SX300.jpg\",\n    \"Metascore\": \"74\",\n    \"imdbRating\": \"8.8\",\n    \"imdbVotes\": \"1,446,708\",\n    \"imdbID\": \"tt1375666\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  },\n  {\n    \"Title\": \"Interstellar\",\n    \"Year\": \"2014\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"07 Nov 2014\",\n    \"Runtime\": \"169 min\",\n    \"Genre\": \"Adventure, Drama, Sci-Fi\",\n    \"Director\": \"Christopher Nolan\",\n    \"Writer\": \"Jonathan Nolan, Christopher Nolan\",\n    \"Actors\": \"Ellen Burstyn, Matthew McConaughey, Mackenzie Foy, John Lithgow\",\n    \"Plot\": \"A team of explorers travel through a wormhole in space in an attempt to ensure humanity's survival.\",\n    \"Language\": \"English\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Won 1 Oscar. Another 39 wins & 132 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BMjIxNTU4MzY4MF5BMl5BanBnXkFtZTgwMzM4ODI3MjE@._V1_SX300.jpg\",\n    \"Metascore\": \"74\",\n    \"imdbRating\": \"8.6\",\n    \"imdbVotes\": \"910,366\",\n    \"imdbID\": \"tt0816692\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  },\n  {\n    \"Title\": \"The Dark Knight\",\n    \"Year\": \"2008\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"18 Jul 2008\",\n    \"Runtime\": \"152 min\",\n    \"Genre\": \"Action, Adventure, Crime\",\n    \"Director\": \"Christopher Nolan\",\n    \"Writer\": \"Jonathan Nolan (screenplay), Christopher Nolan (screenplay), Christopher Nolan (story), David S. Goyer (story), Bob Kane (characters)\",\n    \"Actors\": \"Christian Bale, Heath Ledger, Aaron Eckhart, Michael Caine\",\n    \"Plot\": \"When the menace known as the Joker wreaks havoc and chaos on the people of Gotham, the caped crusader must come to terms with one of the greatest psychological tests of his ability to fight injustice.\",\n    \"Language\": \"English, Mandarin\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Won 2 Oscars. Another 146 wins & 142 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BMTMxNTMwODM0NF5BMl5BanBnXkFtZTcwODAyMTk2Mw@@._V1_SX300.jpg\",\n    \"Metascore\": \"82\",\n    \"imdbRating\": \"9.0\",\n    \"imdbVotes\": \"1,652,832\",\n    \"imdbID\": \"tt0468569\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  },\n  {\n    \"Title\": \"Batman Begins\",\n    \"Year\": \"2005\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"15 Jun 2005\",\n    \"Runtime\": \"140 min\",\n    \"Genre\": \"Action, Adventure\",\n    \"Director\": \"Christopher Nolan\",\n    \"Writer\": \"Bob Kane (characters), David S. Goyer (story), Christopher Nolan (screenplay), David S. Goyer (screenplay)\",\n    \"Actors\": \"Christian Bale, Michael Caine, Liam Neeson, Katie Holmes\",\n    \"Plot\": \"After training with his mentor, Batman begins his fight to free crime-ridden Gotham City from the corruption that Scarecrow and the League of Shadows have cast upon it.\",\n    \"Language\": \"English, Urdu, Mandarin\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Nominated for 1 Oscar. Another 15 wins & 66 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BNTM3OTc0MzM2OV5BMl5BanBnXkFtZTYwNzUwMTI3._V1_SX300.jpg\",\n    \"Metascore\": \"70\",\n    \"imdbRating\": \"8.3\",\n    \"imdbVotes\": \"972,584\",\n    \"imdbID\": \"tt0372784\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  },\n  {\n    \"Title\": \"Avatar\",\n    \"Year\": \"2009\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"18 Dec 2009\",\n    \"Runtime\": \"162 min\",\n    \"Genre\": \"Action, Adventure, Fantasy\",\n    \"Director\": \"James Cameron\",\n    \"Writer\": \"James Cameron\",\n    \"Actors\": \"Sam Worthington, Zoe Saldana, Sigourney Weaver, Stephen Lang\",\n    \"Plot\": \"A paraplegic marine dispatched to the moon Pandora on a unique mission becomes torn between following his orders and protecting the world he feels is his home.\",\n    \"Language\": \"English, Spanish\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Won 3 Oscars. Another 80 wins & 121 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BMTYwOTEwNjAzMl5BMl5BanBnXkFtZTcwODc5MTUwMw@@._V1_SX300.jpg\",\n    \"Metascore\": \"83\",\n    \"imdbRating\": \"7.9\",\n    \"imdbVotes\": \"876,575\",\n    \"imdbID\": \"tt0499549\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  }\n];\n\nfunction getRating(watchList) {\n  // Only change code below this line\n  \n  var averageRating = watchList\n  .filter(e => e.Director === \"Christopher Nolan\")\n  .map(e => Number(e.imdbRating))\n  .reduce((sum, e) => sum + e, 0) / watchList.filter(e => e.Director === \"Christopher Nolan\").length;\n\n  // Only change code above this line\n  return averageRating;\n}\n\nconsole.log(getRating(watchList));",
    "error": null,
    "seed": "// The global variable\nvar watchList = [\n  {\n    \"Title\": \"Inception\",\n    \"Year\": \"2010\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"16 Jul 2010\",\n    \"Runtime\": \"148 min\",\n    \"Genre\": \"Action, Adventure, Crime\",\n    \"Director\": \"Christopher Nolan\",\n    \"Writer\": \"Christopher Nolan\",\n    \"Actors\": \"Leonardo DiCaprio, Joseph Gordon-Levitt, Elliot Page, Tom Hardy\",\n    \"Plot\": \"A thief, who steals corporate secrets through use of dream-sharing technology, is given the inverse task of planting an idea into the mind of a CEO.\",\n    \"Language\": \"English, Japanese, French\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Won 4 Oscars. Another 143 wins & 198 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BMjAxMzY3NjcxNF5BMl5BanBnXkFtZTcwNTI5OTM0Mw@@._V1_SX300.jpg\",\n    \"Metascore\": \"74\",\n    \"imdbRating\": \"8.8\",\n    \"imdbVotes\": \"1,446,708\",\n    \"imdbID\": \"tt1375666\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  },\n  {\n    \"Title\": \"Interstellar\",\n    \"Year\": \"2014\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"07 Nov 2014\",\n    \"Runtime\": \"169 min\",\n    \"Genre\": \"Adventure, Drama, Sci-Fi\",\n    \"Director\": \"Christopher Nolan\",\n    \"Writer\": \"Jonathan Nolan, Christopher Nolan\",\n    \"Actors\": \"Ellen Burstyn, Matthew McConaughey, Mackenzie Foy, John Lithgow\",\n    \"Plot\": \"A team of explorers travel through a wormhole in space in an attempt to ensure humanity's survival.\",\n    \"Language\": \"English\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Won 1 Oscar. Another 39 wins & 132 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BMjIxNTU4MzY4MF5BMl5BanBnXkFtZTgwMzM4ODI3MjE@._V1_SX300.jpg\",\n    \"Metascore\": \"74\",\n    \"imdbRating\": \"8.6\",\n    \"imdbVotes\": \"910,366\",\n    \"imdbID\": \"tt0816692\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  },\n  {\n    \"Title\": \"The Dark Knight\",\n    \"Year\": \"2008\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"18 Jul 2008\",\n    \"Runtime\": \"152 min\",\n    \"Genre\": \"Action, Adventure, Crime\",\n    \"Director\": \"Christopher Nolan\",\n    \"Writer\": \"Jonathan Nolan (screenplay), Christopher Nolan (screenplay), Christopher Nolan (story), David S. Goyer (story), Bob Kane (characters)\",\n    \"Actors\": \"Christian Bale, Heath Ledger, Aaron Eckhart, Michael Caine\",\n    \"Plot\": \"When the menace known as the Joker wreaks havoc and chaos on the people of Gotham, the caped crusader must come to terms with one of the greatest psychological tests of his ability to fight injustice.\",\n    \"Language\": \"English, Mandarin\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Won 2 Oscars. Another 146 wins & 142 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BMTMxNTMwODM0NF5BMl5BanBnXkFtZTcwODAyMTk2Mw@@._V1_SX300.jpg\",\n    \"Metascore\": \"82\",\n    \"imdbRating\": \"9.0\",\n    \"imdbVotes\": \"1,652,832\",\n    \"imdbID\": \"tt0468569\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  },\n  {\n    \"Title\": \"Batman Begins\",\n    \"Year\": \"2005\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"15 Jun 2005\",\n    \"Runtime\": \"140 min\",\n    \"Genre\": \"Action, Adventure\",\n    \"Director\": \"Christopher Nolan\",\n    \"Writer\": \"Bob Kane (characters), David S. Goyer (story), Christopher Nolan (screenplay), David S. Goyer (screenplay)\",\n    \"Actors\": \"Christian Bale, Michael Caine, Liam Neeson, Katie Holmes\",\n    \"Plot\": \"After training with his mentor, Batman begins his fight to free crime-ridden Gotham City from the corruption that Scarecrow and the League of Shadows have cast upon it.\",\n    \"Language\": \"English, Urdu, Mandarin\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Nominated for 1 Oscar. Another 15 wins & 66 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BNTM3OTc0MzM2OV5BMl5BanBnXkFtZTYwNzUwMTI3._V1_SX300.jpg\",\n    \"Metascore\": \"70\",\n    \"imdbRating\": \"8.3\",\n    \"imdbVotes\": \"972,584\",\n    \"imdbID\": \"tt0372784\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  },\n  {\n    \"Title\": \"Avatar\",\n    \"Year\": \"2009\",\n    \"Rated\": \"PG-13\",\n    \"Released\": \"18 Dec 2009\",\n    \"Runtime\": \"162 min\",\n    \"Genre\": \"Action, Adventure, Fantasy\",\n    \"Director\": \"James Cameron\",\n    \"Writer\": \"James Cameron\",\n    \"Actors\": \"Sam Worthington, Zoe Saldana, Sigourney Weaver, Stephen Lang\",\n    \"Plot\": \"A paraplegic marine dispatched to the moon Pandora on a unique mission becomes torn between following his orders and protecting the world he feels is his home.\",\n    \"Language\": \"English, Spanish\",\n    \"Country\": \"USA, UK\",\n    \"Awards\": \"Won 3 Oscars. Another 80 wins & 121 nominations.\",\n    \"Poster\": \"http://ia.media-imdb.com/images/M/MV5BMTYwOTEwNjAzMl5BMl5BanBnXkFtZTcwODc5MTUwMw@@._V1_SX300.jpg\",\n    \"Metascore\": \"83\",\n    \"imdbRating\": \"7.9\",\n    \"imdbVotes\": \"876,575\",\n    \"imdbID\": \"tt0499549\",\n    \"Type\": \"movie\",\n    \"Response\": \"True\"\n  }\n];\n\nfunction getRating(watchList){\n  // Only change code below this line\n  var averageRating;\n\n\n  // Only change code above this line\n  return averageRating;\n}\nconsole.log(getRating(watchList));",
    "editableContents": "",
    "seedEditableRegionBoundaries": [],
    "key": "indexjs"
}
```

**Comment by ojeytonwilliams** [comment]
Do you mean `"key"`?  We can delete that, I think.

File: client/src/templates/Challenges/redux/code-storage-epic.js (hunk 1 above)
Comment (nhcarrigan) [reply client/src/templates/Challenges/redux/code-storage-epic.js:160]: You still want this?

**Review by moT01 — DISMISSED** [review]
This LGTM :tada: 

The only thing I found was that loading code on the beta projects doesn't update the preview. I may have solved this on the PR where I load users code. I would be fine merging with this problem.

Do we still want to drop the `key`?

**Comment by ShaunSHamilton** [comment]
I think we are fine moving forward with this, but one final check, @ojeytonwilliams ?

File: client/src/templates/Challenges/redux/code-storage-epic.js [hunk 3]
```diff
@@ -143,16 +141,44 @@ function loadCodeEpic(action$, state$) {
...
           } else {
             // TODO: After sufficient time, remove parsing of old code-storage format
             // This was pushed to production with https://github.com/freeCodeCamp/freeCodeCamp/pull/43023
             foundChallengeFile = codeFound[challengeFile.fileKey];
+            indexjsCode = codeFound['indexjs'];
+          }
+          let isCodeFound;
+          // Fix the format of the old file
+          if (indexjsCode) {
+            indexjsCode.fileKey = 'scriptjs';
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/redux/code-storage-epic.js:163]:
```suggestion
            indexjsCode.fileKey = 'scriptjs';
            delete indexjsCode.key;
```

**Comment by ojeytonwilliams** [comment]
Probably best to get rid of the `key`, rather than leaving it around when we know it's not necessary.  I've updated the PR.

**Review by ShaunSHamilton — APPROVED** [review]
Thank you, Dr. Oliver!

LGTM 👍

**Review by moT01 — APPROVED** [review]

**Pull request merged**